### PR TITLE
improve release notes generation to look at PR body

### PR DESCRIPTION
**Pull request by shiftkey**
## Overview

Following on from #5994, I worked on refreshing our `draft-release` command to look for the new `Notes` section, and either:

 - uses that text for the release notes header
 - excludes the PR from the list if a `no-notes` value is set

## Description

This is what the output for the last beta release would look like with this PR enabled:

```sh
$ GITHUB_ACCESS_TOKEN=[token] yarn draft-release beta
yarn run v1.10.1
$ ts-node script/draft-release/index.ts beta
Skipping these merged PRs as 'no-notes' was set:
 - #6072 - add support for reporting a binary file is in a conflicted state
 - #6074 - move status tests away from chai
 - #6053 - add some new Git APIs to improve our Git LFS support
 - #6037 - refactor commit message state handling
 - #6044 - styling changes for merge filter text box
 - #6059 - ensure status after aborting a merge reflects reality
 - #6061 - kill off unnecessary margin inside compare tab branch list
 - #6050 - handle really long banner messages
 - #6046 - just some regrets about regexes
 - #6034 - add support for debugging jest tests in vscode

Here's what you should do next:

1. Update the app/package.json 'version' to '1.5.0-beta3' (make sure this aligns with semver format of 'major.minor.patch')
2. Concatenate this to the beginning of the 'releases' element in the changelog.json as a starting point:
{
  "1.5.0-beta3": [
    "[???] Release notes for 1.5 Beta 2 - #6040"
  ]
}

3. Remove any entries of contributions that don't affect the end user
4. Update the release notes to have user-friendly summary lines
5. For issues prefixed with [???], look at the PR to update the prefix to one of: [New], [Added], [Fixed], [Improved], [Removed]
6. Sort the entries so that the prefixes are ordered in this way: [New], [Added], [Fixed], [Improved], [Removed]
7. Commit the changes (on master or as new branch) and push them to GitHub
8. Read this to perform the release: https://github.com/desktop/desktop/blob/master/docs/process/releasing-updates.md
✨  Done in 3.02s.
```

I want to help us fall into two patterns with the work in #5994 

 - simplify the work on the release co-ordinator by having release notes crafted in advance
   - have the discussions about release notes while the PR is evolving, not way after
 - simplify the process of generating release notes when we open PRs like #6083 
   - make it clearer which PRs can be ignored at release time
   - spent fewer cycles iterating on the words and holding up the release

The more I go through these release processes, the more I dislike using the PR title or issue title for the release notes. Why?

 - PR titles are often irrelevant or fun, and I like to keep that spirit around
 - Issue titles often don't end up reflecting what the actual solution was

Got any thoughts on this @desktop/maintainers @desktop/comrades?

## Release notes

Oh yes, and this is not a user-facing change.

Notes: no-notes

### CODE REVIEW

**Comment by say25** [comment]
Would it be easy to make a `Beta Only` property in the PR body that isn't included by default or commented by default to address Beta bug fixes?

File: script/changelog/parser.ts [hunk 1]
```diff
@@ -55,9 +55,34 @@ export function findIssueRef(body: string): string {
...
+  | {
+      readonly kind: 'found' | 'default'
+      readonly text: string
+    }
+  | {
+      readonly kind: 'omitted'
+    }
+
+export function getReleaseNotesDescription(pr: IAPIPR): ReleaseNotes {
+  const re = /[Nn]otes: (.*)/gm
```
Comment (j-f1) [inline script/changelog/parser.ts]: Should this regex start with a `^`? That would stop it from matching `notes:` in the middle of a sentence.

Comment (shiftkey) [reply script/changelog/parser.ts]: 👍 I mostly opened this early to get feedback but I'll throw some better test cases at it while we talk about the goals here.

**Comment by shiftkey** [comment]
@say25 that's actually a great idea. I could add an optional section to the notes to indicate the channel which it was reported on, so we capture the data and can still decide later how to render the release notes.

For example, #6081 could be written like this

```
# Release Notes

Notes: fixes tab order accessibility issue with link in Merge Conflicts dialog
Channel: beta
```

If the channel entry is omitted, `production` is assumed.

**Comment by outofambit** [comment]
i'm into trying this out. i have a little anxiety about asking contributors to write their own release notes as part of the PR submission because it adds a little friction, but totally down to try.

+1 for optional beta channel field

**Comment by shiftkey** [comment]
@outofambit 

> i have a little anxiety about asking contributors to write their own release notes as part of the PR submission because it adds a little friction, but totally down to try.

I'm fine if it gets omitted initially, because reviewers later can come in and flesh it out while they're doing the review. Should we make that more explicit in the comment?

**Comment by outofambit** [comment]
> Should we make that more explicit in the comment?

that sounds good to me!

**Comment by shiftkey** [comment]
Dropping `ready-for-review` because it's been pretty quiet. I plan to update the PR template separate to this and then rewrite this PR to focus on changes to `draft-release` later.

**Pull request closed without merging**